bsi_zoo/cross_val.py

```diff
@@ -0,0 +1,194 @@
+from scipy import linalg
+import numpy as np
+from sklearn.model_selection import GridSearchCV
+from sklearn.base import BaseEstimator, ClassifierMixin, clone
+from sklearn.model_selection import check_cv
+
+
+from .estimators import SpatialSolver
+
+
+def temporal_cv_metric(y, Sigma_Y):
+    """Compute the log-det Bregman divergence between two matrices.
+
+    It is based on the calculation of Gaussian negative log-likelihood.
+    Both matrices needs to be squared and have the same size.
+    It is given by:
+
+    B_inv = inv(B);
+    logdet_distance = trace(A*B_inv) - logdet(A*B_inv) - size(A,1)
+    """
+    # XXX Need improve w.r.t speed (Perhaps use the following conditions)
+    _, logdet = np.linalg.slogdet(Sigma_Y)
+    out = (np.linalg.norm(linalg.sqrtm(np.linalg.inv(Sigma_Y) @ y),
+                          ord='fro')**2 + logdet)
+    return out
+
+
+def _logdet(A):
+    """Compute the log det of a positive semidefinite matrix."""
+    from scipy import linalg
+    vals = linalg.eigvalsh(A)
+    # avoid negative (numerical errors) or zero (semi-definite matrix) values
+    tol = vals.max() * vals.size * np.finfo(np.float64).eps
+    vals = np.where(vals > tol, vals, tol)
+    return np.sum(np.log(vals))
+
+
+def logdet_bregman_div_distance_nll(y, Sigma_Y):
+    """Compute the Gaussian log likelihood of y given covariance Sigma_Y."""
+    # compute empirical covariance of the test set
+    precision = np.inv(Sigma_Y)
+    _, n_features = y.shape
+    log_like = -.5 * (y * (np.dot(y, precision))).sum(axis=1)
+    log_like -= .5 * (n_features * np.log(2. * np.pi) - _logdet(precision))
+    out = np.mean(log_like)
+    return out
```

Comment:
@AliHashemi-ai this comes from mne. Any reason we need something different?

Reply:
@agramfort Initially, I indented to use the MNE code to implement the temporal cross validation metric since the Bregman logdet divergence can be reformulated as negative likelihood loss plus some constants. And since NNL is efficiently implemented in MNE, I wanted to use the code to do it efficiently. But, there was some difficulties that I decided to write the cv cross-validation metric from scratch.